tests/test_iqm_device.py

```diff
@@ -27,3 +31,144 @@ def test_equality_method():
+        decomposed_circuit, circuit, qubit_map={q: q for q in device.qubits[:size]}
+    )
+    print("Circuit qubits:", circuit.all_qubits())
+    print("Routed circuit qubits:", routed_circuit.all_qubits())
+    for q in routed_circuit.all_qubits():
+        if q not in circuit.all_qubits():
+            circuit.append(ops.IdentityGate(1)(q))
+    print("Circuit qubits after adding routing qubits:", circuit.all_qubits())
+    assert_circuits_have_same_unitary_given_final_permutation(routed_circuit, circuit, qubit_map=qubit_map)
+    assert_circuits_have_same_unitary_given_final_permutation(decomposed_routed_circuit, circuit, qubit_map=qubit_map)
+
+
+@pytest.mark.parametrize(
+    "device",
+    ["device_without_resonator", "device_with_resonator", Apollo(), Adonis(), "device_with_multiple_resonators"],
+)
+def test_qubit_connectivity(device: IQMDevice, request):
+    if isinstance(device, str):
+        device = request.getfixturevalue(device)
+    for edge in [(q1, q2) for q1 in device.qubits for q2 in device.qubits if q1 != q2]:
+        gate = ops.CZPowGate()(*edge)
+        if (
+            edge in device.supported_operations[ops.CZPowGate]
+            or tuple(reversed(edge)) in device.supported_operations[ops.CZPowGate]
+        ):
+            device.check_qubit_connectivity(gate)  # This should not raise an error
+        else:
+            with pytest.raises(ValueError):
+                device.check_qubit_connectivity(gate)
+        with pytest.raises(ValueError):
+            device.check_qubit_connectivity(ops.SwapPowGate()(*edge))
+
+
+def test_validate_moves(device_with_resonator):
+    # Test valid MOVE sandwich
+    circuit = Circuit(
+        IQMMoveGate()(device_with_resonator.qubits[0], device_with_resonator.resonators[0]),
+        IQMMoveGate()(device_with_resonator.qubits[0], device_with_resonator.resonators[0]),
+    )
+    assert device_with_resonator.validate_moves(circuit) is None
+    # Test invalid MOVE sandwich
+    circuit = Circuit(
+        IQMMoveGate()(device_with_resonator.qubits[0], device_with_resonator.resonators[0]),
+        IQMMoveGate()(device_with_resonator.qubits[1], device_with_resonator.resonators[0]),
+    )
+    with pytest.raises(ValueError):
+        device_with_resonator.validate_moves(circuit)
+
+    circuit = Circuit(
+        IQMMoveGate()(device_with_resonator.qubits[0], device_with_resonator.resonators[0]),
+        IQMMoveGate()(device_with_resonator.qubits[0], device_with_resonator.qubits[1]),
+    )
+    with pytest.raises(ValueError):
+        device_with_resonator.validate_moves(circuit)
+
+    circuit = Circuit(
+        IQMMoveGate()(device_with_resonator.resonators[0], device_with_resonator.qubits[1]),
+        IQMMoveGate()(device_with_resonator.resonators[0], device_with_resonator.qubits[1]),
+    )
+    with pytest.raises(ValueError):
+        device_with_resonator.validate_moves(circuit)
+
+    # Test valid MOVE without sandwich
```

Comment:
Would it add value to add a test case with one or more sandwiches and a final move that leaves the qubit state in the resonator. That would test the loop at iqm_device.py#367 and the related logic?

Reply:
Yes. I guess I forgot that one.